### PR TITLE
Add support for derivable extended keys for Secp256k1

### PR DESCRIPTION
**Pull request by erasmospunk**
This commit adds an implementation of the BIP32 derivation algorithm that supports "hard" derivations only. In addition to that the bip39 dependency is added for converting mnemonic phrases to extended master key secrets.

Note that NFKD normalisation is disabled in bip39 because of pinned dependencies and it's version should be bumped once it becomes available.

### CODE REVIEW

File: crypto/src/key/secp256k1/extended_keys.rs
Comment (iljakuklic) [inline crypto/src/key/secp256k1/extended_keys.rs:55]: You can use the `.split()` method on `GenericArray` to extract the key part and the chain code part with compile-time bounds checking.

Something like this (may need adjustment):
```rust
use generic_array::{GenericArray, typenum::U32, sequence::Split};
type ByteArray32 = GenericArray<u8, U32>;
let (private_key, chian_code): (ByteArray32, ByteArray32) = result.split();
```

The `ByteArray32` type is going to be convertible to `[u8; 32]` and `&[u8]`.

File: crypto/src/key/secp256k1/extended_keys.rs
Comment (iljakuklic) [inline crypto/src/key/secp256k1/extended_keys.rs:45]: What's the purpose of the `"Bitcoin seed"` string here. Is that for compatibility so that we generate identical key tree to Bitcoin?

File: crypto/src/key/secp256k1/extended_keys.rs
Comment (iljakuklic) [inline crypto/src/key/secp256k1/extended_keys.rs:117]: This part looks very similar to code in `Self::new_master`, could that be factored out?

**Review by iljakuklic — APPROVED** [review]
Looks fine to me, just a couple of comments / suggestions. Please wait for an approval by @muursh or @TheQuantumPhysicist before merging.

File: crypto/src/key/secp256k1/extended_keys.rs
Comment (stanislav-tkach) [inline crypto/src/key/secp256k1/extended_keys.rs:160]: Nitpick: there isn't much sense to use the `test_` prefix.

**Review by stanislav-tkach — APPROVED** [review]

File: crypto/src/key/hdkd/child_number.rs
Comment (TheQuantumPhysicist) [inline crypto/src/key/hdkd/child_number.rs:64]: into, instead of to.

You can also implement the From trait as an alternative solution, which will implement Into for you.

File: crypto/src/key/hdkd/derivation_path.rs
Comment (TheQuantumPhysicist) [reply crypto/src/key/hdkd/derivation_path.rs:34]: I would rather call this join than extend. Both Rust and Python call it join. C++ calls it append.

Also one nice thing boost does is using the operator `/` for these kinds of extensions... it's kind of weird, but shows nicely in code. So if `a`, `b` and `c` are paths, we write in code:

```
let d = a / b / c;
```

In fact, the C++ standard adopted it since C++17: https://en.cppreference.com/w/cpp/filesystem/path/append

I'm not suggesting we go that far, but it's an idea worth thinking about if these joins become annoying or something. If I see it, I won't fight it.

File: crypto/src/key/secp256k1/extended_keys.rs
Comment (TheQuantumPhysicist) [reply crypto/src/key/secp256k1/extended_keys.rs:45]: Yes, it's: https://en.bitcoin.it/wiki/BIP_0032#Master_key_generation

But I'm wondering now whether we want to parametrize this for our own use. Is this a standard that everyone uses in every crypto? Or is it just in bitcoin and everyone adapts it to whatever they want?

File: crypto/src/key/secp256k1/extended_keys.rs
Comment (TheQuantumPhysicist) [inline crypto/src/key/secp256k1/extended_keys.rs:110]: We usually follow the format TODO(TAG). It makes it easier to search for all tagged TODOs.

File: crypto/src/key/hdkd/child_number.rs
Comment (TheQuantumPhysicist) [reply crypto/src/key/hdkd/child_number.rs:64]: Btw, same applies to the function above

File: crypto/src/key/secp256k1/extended_keys.rs
Comment (TheQuantumPhysicist) [inline crypto/src/key/secp256k1/extended_keys.rs:101]: Isn't this unreachable from ChildNumber itself, guaranteed from the type? Why are we checking for it here again?

File: crypto/src/key/secp256k1/extended_keys.rs
Comment (TheQuantumPhysicist) [inline crypto/src/key/secp256k1/extended_keys.rs:105]: Please move all these stateful mac updates into a separate function. The rule of thumb is: Whenever we play with `mut`, we try to isolate it. It's not always possible, but we do it whenever we can, and it seems to be possible in your case.

So, do the check if `num` is hardened (which I'm not convinced is necessary, since like I said, it's not supported anyway at the `ChildNumber` level). I don't know why we're checking then appending `0u8` followed by `secp_key`. Does being hardened have any bearing on that? If yes, it would be nice to move any specifical part of "being hardened" into the `ChildNumber` struct, and just retrieve it with a method. I think we can at least take the `secp_key` appending out as it's neutral and doesn't matter.

Then what do we have left? We create a function that takes a ref to the private key data + a ref to num, and returns the result array. That would be beautifully more readable and separately testable.

File: crypto/src/key/secp256k1/extended_keys.rs
Comment (TheQuantumPhysicist) [inline crypto/src/key/secp256k1/extended_keys.rs:137]: Please call this function `public_key()`. It's a simple getter.

File: crypto/src/key/mod.rs
Comment (TheQuantumPhysicist) [inline crypto/src/key/mod.rs:51]: I don't know why we're doing this. These don't have to be mutually exclusive. In fact, it's not really possible to encode/decode one into the other. Usually, when you decode bytes into whole objects, you use `decode_all`, which will guarantee that the whole object is filled.

Please don't use this unnecessary and confusing limitation. The only relationship between KeyKind and ExtendedKeyKind is that ExtendedKey can be converted into Key. There's no other hidden relationships here. We don't want the reader of the code to be solving puzzles.

File: crypto/src/key/mod.rs
Comment (TheQuantumPhysicist) [inline crypto/src/key/mod.rs:59]: Based on the discussion above, this has to be `indes = 0`

File: crypto/src/key/mod.rs
Comment (TheQuantumPhysicist) [inline crypto/src/key/mod.rs:166]: Shouldn't there be a method that extracts the private key from the extended private key? How else are we gonna sign stuff?

File: crypto/src/key/mod.rs
Comment (TheQuantumPhysicist) [inline crypto/src/key/mod.rs:205]: Same... where's the conversion function?

File: crypto/src/key/mod.rs
Comment (TheQuantumPhysicist) [inline crypto/src/key/mod.rs:73]: Please move all these new things into an `extended.rs` module. You're also free to move everything else and make this file just for `mod` and `use` statements. We try everything we can to avoid clutter in code.

File: crypto/src/key/mod.rs
Comment (TheQuantumPhysicist) [inline crypto/src/key/mod.rs:225]: This trait impl shouldn't exist. The struct `Secp256k1ExtendedPrivateKey` is not publicly accessible, and this is only used in new. So to save 20 more characters in the `new` function(s), we're adding a whole trait impl. Not a good trade-off. 

Same applies to `Secp256k1ExtendedPublicKey`.

File: crypto/src/key/mod.rs
Comment (TheQuantumPhysicist) [inline crypto/src/key/mod.rs:339]: derive_secp256k1schnorr

**Review by TheQuantumPhysicist — APPROVED** [review]
Looks good and fulfills its purpose. There are many little things that have to be fixed. Please update them and request another review from me. Since this is the first time we're adding extended private keys, I'm gonna be a little nit-picky in this :-)

File: crypto/src/key/secp256k1/extended_keys.rs
Comment (iljakuklic) [reply crypto/src/key/secp256k1/extended_keys.rs:45]: The second hardened component is usually a coin type in the usual layout. That way you can have a wallet that communicates with multiple networks and is able to generate keys for all from the same seed. Since the Mintlayer wallet is supposed to interface both Mintlayer and Bitcoin, I think we should just use the `"Bitcoin seed"` string and allocate a new number for the second derivation path component. Unless there are other complications I have not considered, that is.

https://learnmeabitcoin.com/technical/derivation-paths#wallet-structure

File: crypto/src/key/hdkd/derivation_path.rs
Comment (iljakuklic) [reply crypto/src/key/hdkd/derivation_path.rs:34]: Join and append differ in that the RHS can have multiple components whereas the method presented here only adds one component at the end. In that respect, it's more similar to `Vec::push`. If it was mutating the argument in place, it would make sense to just call it `push`.

Since it is internally and conceptually a `Vec`, it makes sense to me for the API to mimic `Vec`. Maybe it is sensible to change this method to use `&mut self` and call it `push`. If you look how it's implemented, there is hidden clone followed by a push anyway. For completenes can also add `FromIterator` to add multiple child numbers in one go and `join` to concatenate two paths together. I think this would make the API largely unsurprising.

Comment (TheQuantumPhysicist) [reply crypto/src/key/hdkd/derivation_path.rs:34]: Structurally, it's a path. But it's not implemented like a path in the API for some reason. If I understand correctly, you're not happy calling this function a join because the API we created doesn't act like a path. Well, I'd call that a problem in the API, not in the idea, since this structure is basically a path in every way. I hope I'm not missing something in this story.

And just to not give the vibe that I'm insistent, I wanna emphasize that I don't care too much about this detail since this is some esoteric thing that'll barely be used in the code. But it'd be nice to have something consistent coming out of it.

File: crypto/src/key/mod.rs
Comment (erasmospunk) [reply crypto/src/key/mod.rs]: I have reverted this have the same index for both extended and normal keys. The rationale is that by having a different "version" number it is clear which key is which.

File: crypto/src/key/hdkd/derivation_path.rs
Comment (erasmospunk) [reply crypto/src/key/hdkd/derivation_path.rs]: I have changed the API by adding a push & pop operation.

File: crypto/src/key/secp256k1/extended_keys.rs
Comment (erasmospunk) [reply crypto/src/key/secp256k1/extended_keys.rs]: The "Bitcoin seed" constant is defined in BIP39 and it is used for most of the coins. Some coins do change it but it is when they use different elliptic curve. I our case we don't need to change it.

File: crypto/src/key/secp256k1/extended_keys.rs
Comment (erasmospunk) [reply crypto/src/key/secp256k1/extended_keys.rs]: Good point, I changed it. I have searched and there are many tests that are `fn test_`

File: crypto/src/key/secp256k1/extended_keys.rs
Comment (erasmospunk) [reply crypto/src/key/secp256k1/extended_keys.rs]: Removed this

File: crypto/src/key/secp256k1/extended_keys.rs
Comment (erasmospunk) [reply crypto/src/key/secp256k1/extended_keys.rs]: I moved the some of the functions to the `to_key_and_chain_code` function. I tried moving out the mut parts to a separate function but it makes the code harder to follow because it breaks in half the BIP32 algorithm.

I simplified the ChildNumber API so we don't have to deal with non-hardened paths. The `0u8` is needed for the private key padding and it is unrelated to the child number.

File: crypto/src/key/mod.rs
Comment (erasmospunk) [reply crypto/src/key/mod.rs]: This was a mistake. I have added a `private_key()` method. Unfortunately I couldn't find a way to make it return a reference.

Comment (iljakuklic) [reply crypto/src/key/mod.rs]: > This was a mistake. I have added a `private_key()` method. Unfortunately I couldn't find a way to make it return a reference.

I don't think it's possible to return the final private key as a reference unless you pre-compute the key and store it with in the struct. Otherwise, you get a reference to a local variable which is not allowed. In that case, I'd say just return it by value.

**Review by iljakuklic — APPROVED** [review]
The updated version looks good to me

File: crypto/src/key/mod.rs
Comment (TheQuantumPhysicist) [reply crypto/src/key/mod.rs]: While that sounds nice, the problem becomes consistency in the future if we ever add something else. The size should be enough, even though it's not perfect in a long stream.

**Review by TheQuantumPhysicist — APPROVED** [review]
All good! Cheers!

**Pull request merged**
